waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SaslThriftMetastoreClientManager.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.hadoop.util.StringUtils;
+import org.apache.hive.service.auth.KerberosSaslHelper;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransport;
+
+import lombok.extern.log4j.Log4j2;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import com.hotels.bdp.waggledance.client.compatibility.HiveCompatibleThriftHiveMetastoreIfaceFactory;
+import com.hotels.bdp.waggledance.context.CommonBeans;
+
+@Log4j2
+public class SaslThriftMetastoreClientManager extends AbstractThriftMetastoreClientManager {
+
+  private final boolean impersonationEnabled;
+  private static final Duration delegationTokenCacheTtl = Duration.ofHours(
+      1); // The default lifetime in Hive is 7 days (metastore.cluster.delegation.token.max-lifetime)
+  private static final long delegationTokenCacheMaximumSize = 1000;
+  private static final LoadingCache<DelegationTokenKey, String> delegationTokenCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(delegationTokenCacheTtl.toMillis(), MILLISECONDS)
+      .maximumSize(delegationTokenCacheMaximumSize)
+      .build(CacheLoader.from(SaslThriftMetastoreClientManager::loadDelegationToken));
+
+  SaslThriftMetastoreClientManager(HiveConf conf,
+      HiveCompatibleThriftHiveMetastoreIfaceFactory hiveCompatibleThriftHiveMetastoreIfaceFactory,
+      int connectionTimeout) {
+    super(conf, hiveCompatibleThriftHiveMetastoreIfaceFactory, connectionTimeout);
+    impersonationEnabled = conf.getBoolean(CommonBeans.IMPERSONATION_ENABLED_KEY, false);
+  }
+
+  @Override
+  void open(HiveUgiArgs ugiArgs) {
+    if (isConnected) {
+      return;
+    }
+    createMetastoreClientAndOpen(null, ugiArgs);
+    if (impersonationEnabled) {
+      try {
+        String userName = UserGroupInformation.getCurrentUser().getShortUserName();
+        DelegationTokenKey key = new DelegationTokenKey(msUri, userName, client);
+        String delegationToken = delegationTokenCache.get(key);
+        close();
+        createMetastoreClientAndOpen(delegationToken, ugiArgs);
+      } catch (IOException | ExecutionException e) {
+        log.error("Couldn't create delegation token client");
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  void createMetastoreClientAndOpen(String delegationToken, HiveUgiArgs ugiArgs) {
+    TException te = null;
+    boolean useSsl = conf.getBoolVar(ConfVars.HIVE_METASTORE_USE_SSL);
+    boolean useCompactProtocol = conf.getBoolVar(ConfVars.METASTORE_USE_THRIFT_COMPACT_PROTOCOL);
+    int clientSocketTimeout = (int) conf.getTimeVar(ConfVars.METASTORE_CLIENT_SOCKET_TIMEOUT,
+        TimeUnit.MILLISECONDS);
+
+    for (int attempt = 0; !isConnected && (attempt < retries); ++attempt) {
```

Comment:
I would recommend splitting this section into smaller private method and/or helper classes. The comments alone should be a good indication that it's hard to understand what's going on and why.

Reply:
Fair enough. @patduin explained to me where this is coming from. It's better to leave it as is. Thanks.